wpengine/atlas-content-modeler
# [MTKA-1319] Fix relationship cardinality restrictions on the reverse side

## Description

A final fix to make reverse relationships shippable by flipping cardinality on the reverse side so that it's enforced correctly when editing back references:

- One-to-many fields now appear as a many-to-one on the reverse side.
- Many-to-one fields now appear as a one-to-many on the reverse side.
- If you’re editing a one-to-one from the B side, you can only change to other As that aren’t connected to Bs (no changes were required for this — the PR just adds a test to check it's working).
- If you’re editing a many-to-one from the B side (where it now appears as a one-to-many), you can only select new As that don’t already have a connection with another B (no changes required to JS logic, flipping cardinality was enough).

Also updates the label for many-to-one fields so that the button reads “Link Right” instead of
“Link Rights” (only one item is selectable for a many-to-one field, so the label should be singular not plural).

https://wpengine.atlassian.net/browse/MTKA-1319

## Testing

- Includes a PHP unit test to check that cardinality is reversed.
- Includes an e2e test to check that cardinality is enforced from the reverse side via the UI.

### Manual testing

Import these models (you'll need to remove `.txt` from the file name).

[acm-models-export-2021-11-15-11-20-09.json.txt](https://github.com/wpengine/atlas-content-modeler/files/7538244/acm-models-export-2021-11-15-11-20-09.json.txt)

1. Create two “Left” posts named Left 1 and Left 2, filling only the Name field.
2. Create a new “Right 1” post, linking the “Left One to One” field to “Left 1” the “Left One To Many” field to both Lefts.
3. Start creating a new “Right 2” post.

In the Right 2 post, confirm that:

- “Left 1” is not selectable in the “Left One to One” field. (It is already linked to Right 1.)
- Neither “Left 1” nor “Left 2” are selectable in the “Left One To Many” field. (Both are already linked to Right 1.)

## Screenshots



## Documentation Changes

Cardinality will be documented separately on the [dev docs site](https://developers.wpengine.com/docs/atlas-content-modeler/introduction).

## Dependant PRs

Targets #337 and should be merged here first before final review there.
